lib/iris/_data_manager.py

```diff
@@ -187,6 +187,13 @@ def _deepcopy(self, memo, data=None):
                     data = copy.deepcopy(self._lazy_array, memo)
                 else:
                     data = self._real_array.copy()
+            else:
+                # Check that the replacement data is valid relative to
+                # the currently managed data.
+                dm_check = DataManager(self.core_data())
+                dm_check.data = data
+                # If the replacement data is valid, then use it but
+                # without copying it.
```

Comment:
I am putting this back as this ensures you are doing a deepcopy with a data of the correct shape. 
Without this you can do:
```
>>> dm = DataManager(np.array([0,1,2]))
>>> dm._deepcopy(dict(), data=np.array(0))
DataManager(array(0))
```

Which shouldn't happen